hcl/cty/object_test.go

```diff
@@ -0,0 +1,181 @@
+// Copyright 2022 Mineiros GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cty_test
+
+import (
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/mineiros-io/terramate/errors"
+	"github.com/mineiros-io/terramate/hcl/cty"
+	errtest "github.com/mineiros-io/terramate/test/errors"
+)
+
+func TestCtyObjectSetAt(t *testing.T) {
+	type testcase struct {
+		name    string
+		obj     *cty.Object
+		val     interface{}
+		path    string
+		want    *cty.Object
+		wantErr error
+	}
+
+	for _, tc := range []testcase{
+		{
+			name: "set at root, empty object",
+			obj:  cty.NewObject(),
+			path: "key",
+			val:  "value",
+			want: &cty.Object{
+				Keys: map[string]interface{}{
+					"key": "value",
+				},
+			},
+		},
+		{
+			name: "set at root, override value",
+			obj: &cty.Object{
+				Keys: map[string]interface{}{
+					"key": "old value",
+				},
+			},
+			path: "key",
+			val:  "new value",
+			want: &cty.Object{
+				Keys: map[string]interface{}{
+					"key": "new value",
+				},
+			},
+		},
+		{
+			name: "set at root, new value",
+			obj: &cty.Object{
+				Keys: map[string]interface{}{
+					"key": "value",
+				},
+			},
+			path: "other-key",
+			val:  "other value",
+			want: &cty.Object{
+				Keys: map[string]interface{}{
+					"key":       "value",
+					"other-key": "other value",
+				},
+			},
+		},
+		{
+			name: "set at an existing child object",
+			obj: &cty.Object{
+				Keys: map[string]interface{}{
+					"key": cty.NewObject(),
+				},
+			},
+			path: "key.test",
+			val:  "child value",
+			want: &cty.Object{
+				Keys: map[string]interface{}{
+					"key": &cty.Object{
+						Keys: map[string]interface{}{
+							"test": "child value",
+						},
+					},
+				},
+			},
+		},
```

Comment:
```suggestion
		{
			name: "set at an existing child object",
			obj: &cty.Object{
				Keys: map[string]any{
					"key": cty.NewObject(),
				},
			},
			path: "key.test",
			val:  "child value",
			want: &cty.Object{
				Keys: map[string]any{
					"key": &cty.Object{
						Keys: map[string]any{
							"test": "child value",
						},
					},
				},
			},
		},
```

Can also do in other tests, it gets a little shorter.